OSDOCS-12906: Linked HTTP/2 doc in route config guide (#87145)
openshift/openshift-docs: modules/nw-http2-haproxy.adoc

//
// * networking/ingress-operator.adoc

:_mod-docs-content-type: CONCEPT
[id="nw-http2-haproxy_{context}"]
= Enable and disable HTTP/2 on Ingress Controllers

You can enable or disable transparent end-to-end HTTP/2 connectivity in HAProxy. This allows application owners to make use of HTTP/2 protocol capabilities, including single connection, header compression, binary streams, and more.

You can enable or disable HTTP/2 connectivity for an individual Ingress Controller or for the entire cluster.

[NOTE]
====

> If you enable HTTP/2 connectivity for an individual Ingress Controller and for the entire cluster, the HTTP/2 configuration for the Ingress Controller takes precedence over the HTTP/2 configuration for the cluster. 

Author: :heavy_check_mark:

====

> Thus configuring HTTP/2 on the IngressController and configuring it on the cluster config are not mutually exclusive options, contrary to what the "either" could imply.

Author: Removed. Thanks!

To enable the use of HTTP/2 for a connection from the client to an HAProxy instance, a route must specify a custom certificate. A route that uses the default certificate cannot use HTTP/2. This restriction is necessary to avoid problems from connection coalescing, where the client re-uses a connection for different routes that use the same certificate.

[NOTE]
====
For an HAProxy instance that runs on a back-end pod, you can set `appProtocol: kubernetes.io/h2c` in the service specification, so that the router can read the service specification and then send traffic to a target application. For security reasons, you cannot apply this configuration for an HAProxy instance that runs on a front-end pod.
====

> The connection from HAProxy to the application pod can use HTTP/2 only for re-encrypt routes and not for edge-terminated or insecure routes. This restriction is because HAProxy uses Application-Level Protocol Negotiation (ALPN), which is a TLS extension, to negotiate the use of HTTP/2 with the back-end. The implication is that end-to-end HTTP/2 is possible with passthrough and re-encrypt and not with edge-terminated routes.

Author: :heavy_check_mark:

//
// * networking/ingress-operator.adoc

:_mod-docs-content-type: CONCEPT
[id="nw-http2-haproxy_{context}"]
= Enable and disable HTTP/2 on Ingress Controllers

You can enable or disable transparent end-to-end HTTP/2 connectivity in HAProxy. This allows application owners to make use of HTTP/2 protocol capabilities, including single connection, header compression, binary streams, and more.

You can enable or disable HTTP/2 connectivity for an individual Ingress Controller or for the entire cluster.

[NOTE]
====
If you enable or disable HTTP/2 connectivity for an individual Ingress Controller and for the entire cluster, the HTTP/2 configuration for the Ingress Controller takes precedence over the HTTP/2 configuration for the cluster. 
====

To enable the use of HTTP/2 for a connection from the client to an HAProxy instance, a route must specify a custom certificate. A route that uses the default certificate cannot use HTTP/2. This restriction is necessary to avoid problems from connection coalescing, where the client re-uses a connection for different routes that use the same certificate.

Consider the following use cases for an HTTP/2 connection for each route type: 

> * For a re-encrypt route, the conenction from HAProxy to the application pod can use HTTP/2 if the application supports using Application-Level Protocol Negotiation (ALPN) to negotiate HTTP/2 with HAProxy. You cannot use HTTP/2 with a re-encrypt route unless the Ingress Controller has HTTP/2 enabled.

Reviewer: Typo: "conenction".  (That's actually a typo in my earlier suggestion, sorry.)

Author: Hi @Miciah . NP. Thanks. Fixed!

> * For a re-encrypt route, the conenction from HAProxy to the application pod can use HTTP/2 if the application supports using Application-Level Protocol Negotiation (ALPN) to negotiate HTTP/2 with HAProxy.
> * For a passthrough route, the connection can use HTTP/2 if the application supports using ALPN to negotiate HTTP/2 with the client.
> * For an edge-terminated secure route, the connection uses HTTP/2 if the service specifies only `appProtocol: kubernetes.io/h2c`.
> * For an insecure route, the connection uses HTTP/2 if the service specifies only `appProtocol: kubernetes.io/h2c`. You can use HTTP/2 with an insecure route if the Ingress Controller has HTTP/2 enabled or disabled.

Reviewer: The word "only" is unnecessary.  (I wrote "if (and only if)" earlier, and that would be clear, but "if ... only" seems confusing to me.)

The last statement, "You can use HTTP/2 with an insecure route if the Ingress Controller has HTTP/2 enabled or disabled", applies to passthrough TLS and edge-terminated secure routes as well.  

```suggestion
* For a re-encrypt route, the conenction from HAProxy to the application pod can use HTTP/2 if the application supports using Application-Level Protocol Negotiation (ALPN) to negotiate HTTP/2 with HAProxy. You cannot use HTTP/2 with a re-encrypt route unless the Ingress Controller has HTTP/2 enabled.
* For a passthrough route, the connection can use HTTP/2 if the application supports using ALPN to negotiate HTTP/2 with the client. You can use HTTP/2 with a passthrough route if the Ingress Controller has HTTP/2 enabled or disabled.
* For an edge-terminated secure route, the connection uses HTTP/2 if the service specifies `appProtocol: kubernetes.io/h2c`. You can use HTTP/2 with an edge-terminated secure route if the Ingress Controller has HTTP/2 enabled or disabled.
* For an insecure route, the connection uses HTTP/2 if the service specifies `appProtocol: kubernetes.io/h2c`. You can use HTTP/2 with an insecure route if the Ingress Controller has HTTP/2 enabled or disabled.
```

Author: Thanks, updated!

[id="nw-http2-haproxy_{context}"]
= Enabling HTTP/2 Ingress connectivity

You can enable transparent end-to-end HTTP/2 connectivity in HAProxy. By enabling HTTP/2, an application owner can access HTTP/2 protocol capabilities, which includes single connection, header compression, binary streams, and other features.

You can enable HTTP/2 connectivity for an individual Ingress Controller or for the entire cluster.

To enable the use of HTTP/2 for the connection from the client to HAProxy, a route must specify a custom certificate. A route that uses the default certificate cannot use HTTP/2. This restriction is necessary to avoid problems from connection coalescing, where the client re-uses a connection for different routes that use the same certificate.

[NOTE]
====

> For a service that runs on a back-end pod, you can set `appProtocol: kubernetes.io/h2c` in the service configuration file, so that the service can send traffic over cleartext routes. You cannot apply this configuration for a service that runs on a front-end pod. 

Author: Is this considered safe? The Cleartext TCP seems to sending data over a route with no encryption or encoding of any kind.

> For a service that runs on a back-end pod, you can set `appProtocol: kubernetes.io/h2c` in the service configuration file, so that the service can send traffic over cleartext routes. You cannot apply this configuration for a service that runs on a front-end pod. 

Reviewer: I disagree.  "Cleartext", meaning "not encrypted", is the proper term for this context.  "Plain text" means human-readable (typically ASCII) as opposed to binary-encoded data.  I cannot find anything in the IBM Style Guide to justify writing "plain text" instead of "cleartext" here.

[id="nw-http2-haproxy_{context}"]
= Enabling HTTP/2 Ingress connectivity

> You can enable transparent end-to-end HTTP/2 connectivity in HAProxy, so that an application owner can access HTTP/2 protocol capabilities, which includes single connection, header compression, binary streams, and other features.

Reviewer: I'm not sure that it is useful to list capabilities.  If I didn't already know what you meant by "single connection", I would not understand from this statement, and I might even infer something false.  

(HTTP/1.1 supports pipelining, which allows multiple serial requests over a single TCP connection.  HTTP/2 adds multiplexing of multiple concurrent requests over a single TCP connection.  I'm guessing that "single connection" is referring to the latter.)

Author: Updated to "such as multiplexing of multiple concurrent requests over a single Transmission Control Protocol (TCP) connection."

> You can enable transparent end-to-end HTTP/2 connectivity in HAProxy, so that an application owner can access HTTP/2 protocol capabilities, which includes single connection, header compression, binary streams, and other features.

Reviewer 2: ```suggestion
You can enable transparent end-to-end HTTP/2 connectivity in HAProxy, so that an application owner can access HTTP/2 protocol capabilities, which include single connection, header compression, binary streams, and other features.
```
You could also use "such as" instead of "which include".

> You can either enable HTTP/2 connectivity for an individual Ingress Controller or for the entire cluster.

Reviewer: To elaborate on this a little, you can specify an explicit true/false preference on an IngressController, and that takes precedence over any true/false annotation on the cluster config.  Thus configuring HTTP/2 on the IngressController and configuring it on the cluster config are not mutually exclusive options, contrary to what the "either" could imply.

Author: Added:

> [NOTE]
====
If you enable HTTP/2 connectivity for an individual Ingress Controller and for the entire cluster, the HTTP/2 configuration for the Ingress Controller takes precedence over the HTTP/2 configuration for the cluster. 
====

> To enable the use of HTTP/2 for a connection from the client to a HAProxy service, a route must specify a custom certificate. A route that uses the default certificate cannot use HTTP/2. This restriction is necessary to avoid problems from connection coalescing, where the client re-uses a connection for different routes that use the same certificate.

Reviewer: We use "an HAProxy" rather than "a HAProxy" elsewhere in the documentation.  It is a stylistic choice that depends on how one pronounces the 'h', but "an HAProxy" is more consistent with existing documentation.

In any case, why write "a(n) HAProxy service" instead of simply HAProxy?  In my opinion, it is clearer to write "HAProxy", "an HAProxy instance", or "a router pod".  The word "service" could imply that a Kubernetes service is involved, which is not always the case.

[NOTE]
====

> For a HAProxy service that runs on a back-end pod, you can set `appProtocol: kubernetes.io/h2c` in the service configuration file, so that the service can send traffic over an unencrypted route. For security reasons, you cannot apply this configuration for a HAProxy service that runs on a front-end pod.

Reviewer: The phrase "the service can send traffic over an unencrypted route" is confusing.  It's the router (that is, HAProxy) which sends traffic to the application, and it doesn't send it over the service; the router uses the service spec to read configuration, but the router generally does not send traffic to the service's cluster IP address.  

I don't know what a "a front-end pod" means or what restriction this sentence is referring to: "For security reasons, you cannot apply this configuration for a HAProxy service that runs on a front-end pod."

> For a HAProxy service that runs on a back-end pod, you can set `appProtocol: kubernetes.io/h2c` in the service configuration file, so that the service can send traffic over an unencrypted route. For security reasons, you cannot apply this configuration for a HAProxy service that runs on a front-end pod.

Reviewer 2: ```suggestion
For an HAProxy service that runs on a back-end pod, you can set `appProtocol: kubernetes.io/h2c` in the service configuration file so that the service can send traffic over an unencrypted route. For security reasons, you cannot apply this configuration for an HAProxy service that runs on a front-end pod.
```

> For a HAProxy service that runs on a back-end pod, you can set `appProtocol: kubernetes.io/h2c` in the service configuration file, so that the service can send traffic over an unencrypted route. For security reasons, you cannot apply this configuration for a HAProxy service that runs on a front-end pod.

Author: Updated to state:

> For an HAProxy instance that runs on a back-end pod, you can set `appProtocol: kubernetes.io/h2c` in the service specification, so that the router can read the service specification and then send traffic to a target application.

Front-end pod = a customer-facing application. Drop the "For security reasons, you cannot apply this configuration for an HAProxy instance that runs on a front-end pod." sentence?

> For a HAProxy service that runs on a back-end pod, you can set `appProtocol: kubernetes.io/h2c` in the service configuration file, so that the service can send traffic over an unencrypted route. For security reasons, you cannot apply this configuration for a HAProxy service that runs on a front-end pod.

Reviewer: Yeah, it's still confusing.  I'd drop the sentence about front-end pods, and drop the phrase about backend-pods:
```yaml
You can set `appProtocol: kubernetes.io/h2c` in the service specification to specify that the router should use clear-text HTTP/2 to send traffic to a target application.
```

====

The connection from HAProxy to the application pod can use HTTP/2 only for re-encrypt routes and not for edge-terminated or insecure routes. This restriction is because HAProxy uses Application-Level Protocol Negotiation (ALPN), which is a TLS extension, to negotiate the use of HTTP/2 with the back-end. The implication is that end-to-end HTTP/2 is possible with passthrough and re-encrypt and not with insecure or edge-terminated routes.

[IMPORTANT]
====

> For non-passthrough routes, the Ingress Controller negotiates its connection to the application independently of the connection from the client. This means a client might connect to the Ingress Controller and negotiate HTTP/1.1; in turn, the Ingress Controller might connect to the application, negotiate HTTP/2, and forward the request from the client HTTP/1.1 connection by using the HTTP/2 connection to the application. This situation poses a problem if the client subsequently tries to upgrade its connection from HTTP/1.1 to the WebSocket protocol, because the Ingress Controller cannot forward WebSocket to HTTP/2 and cannot upgrade its HTTP/2 connection to WebSocket. 

Reviewer 2: ```suggestion
For non-passthrough routes, the Ingress Controller negotiates its connection to the application independently of the connection from the client. This means a client might connect to the Ingress Controller and negotiate HTTP/1.1. In turn, the Ingress Controller might connect to the application, negotiate HTTP/2, and forward the request from the client HTTP/1.1 connection by using the HTTP/2 connection to the application. This situation poses a problem if the client subsequently tries to upgrade its connection from HTTP/1.1 to the WebSocket protocol because the Ingress Controller cannot forward WebSocket to HTTP/2 and cannot upgrade its HTTP/2 connection to WebSocket. 
```

[IMPORTANT]
====
For non-passthrough routes, the Ingress Controller negotiates its connection to the application independently of the connection from the client. This means a client might connect to the Ingress Controller and negotiate HTTP/1.1. In turn, the Ingress Controller might connect to the application, negotiate HTTP/2, and forward the request from the client HTTP/1.1 connection by using the HTTP/2 connection to the application. This situation poses a problem if the client subsequently tries to upgrade its connection from HTTP/1.1 to the WebSocket protocol, because the Ingress Controller cannot forward WebSocket to HTTP/2 and cannot upgrade its HTTP/2 connection to WebSocket. 

> Consequently, if you have an application that is intending to accept WebSocket connections, and the application attempts to allow for HTTP/2 protocol negotiation, the client fails to upgrade to the WebSocket protocol.

Author: :heavy_check_mark:
